### PR TITLE
Refactor Association into Field like everything else

### PR DESCRIPTION
**Pull request by bf4**
Intended to simplify work in https://github.com/rails-api/active_model_serializers/pull/1857

### CODE REVIEW

**Comment by bf4** [comment]
@NullVoxPopuli Where do you think a regression test for something like https://github.com/rails-api/active_model_serializers/pull/1857#issuecomment-243643519 should go?

**Comment by NullVoxPopuli** [comment]
@bf4, maybe in `test/active_model_serializers/adapter/json_api/{type_test.rb, belongs_to_test.rb}`?

It looks like both test files don't do anything with polymorphic belongs_to

File: lib/active_model/serializer/association.rb [hunk 1]
```diff
@@ -3,17 +3,32 @@ class Serializer
...
-    # @attr [ActiveModel::Serializer] serializer
     # @attr [Hash{Symbol => Object}] options
+    # @attr [block]
     #
     # @example
-    #  Association.new(:comments, CommentSummarySerializer)
+    #  Association.new(:comments, { serializer: CommentSummarySerializer })
     #
-    Association = Struct.new(:name, :serializer, :options, :links, :meta) do
+    class Association < Field
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/association.rb]: :+1:

File: lib/active_model/serializer/associations.rb [hunk 2]
```diff
@@ -74,8 +74,8 @@ def has_one(name, options = {}, &block) # rubocop:disable Style/PredicateName
         # @api private
         #
         def associate(reflection)
-          key = reflection.options[:key]
-          key ? self._reflections[key] = reflection : self._reflections[reflection.name] = reflection
+          key = reflection.options[:key] || reflection.name
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: Isn't there a rubocop setting to yell at us for the previous style?

File: lib/active_model/serializer/reflection.rb [hunk 3]
```diff
@@ -106,17 +106,19 @@ def value(serializer)
       #
       # @api private
       #
-      def build_association(subject, parent_serializer_options)
-        association_value = value(subject)
+      def build_association(parent_serializer, parent_serializer_options)
+        association_value = value(parent_serializer)
         reflection_options = options.dup
-        serializer_class = subject.class.serializer_for(association_value, reflection_options)
+        serializer_class = parent_serializer.class.serializer_for(association_value, reflection_options)
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: for just api-design, should a parent_serializer instance have a serialize_for method that just delegates to the class? would that semantically be weird?

**Comment by NullVoxPopuli** [comment]
I'm a fan overall.
Also, looks like git can't figure out that files moving (#1898) doesn't mean there _has_ to be conflicts

File: lib/active_model/serializer/reflection.rb (hunk 3 above)
Comment (bf4) [inline lib/active_model/serializer/reflection.rb]: I sort of don't think it should a method on the serializer class in the first place.. but I've also seen people override the method per serializer.  dunno

**Comment by bf4** [comment]
@NullVoxPopuli rebased

https://github.com/rails-api/active_model_serializers/pull/1897#issuecomment-243730331

I"m not sure what your comment is referring to.  (The file is `test/adapter/json_api/` btw)

**Comment by NullVoxPopuli** [comment]
> I"m not sure what your comment is referring to. (The file is test/adapter/json_api/ btw)

you asked:

> @NullVoxPopuli Where do you think a regression test for something like #1857 (comment) should go?

? or did you mean something else?

**Comment by bf4** [comment]
Looks like the inflection testing in test/adapter/json_api/type_test.rb is really testing `lib/active_model_serializers/adapter/json_api/resource_identifier.rb`

**Comment by NullVoxPopuli** [comment]
> Looks like the inflection testing in test/adapter/json_api/type_test.rb is really testing lib/active_model_serializers/adapter/json_api/resource_identifier.rb

That should probably move to the test file for resource_identifier then.

**Comment by bf4** [comment]
@NullVoxPopuli oh, github presentation of doc led me not to see your comment as a response to mine :facepalm:

**Comment by bf4** [comment]
I'm not planning any more changes for this PR. Anything you suggest?

**Comment by NullVoxPopuli** [comment]
Nope, this is a good chunk of _good_ changes :-)

I like when things get less convoluted :-)

I'll merge when CI passes

**Pull request merged**